src/components/encrypted-intent.ts

```diff
@@ -0,0 +1,237 @@
+
+/**
+ * Implements some special handling on top of Intent to handle encrypted rooms.
+ */
+export class EncryptedIntent extends Intent {
+
+    private readonly encryptedRooms = new Map<string, string|false>();
+    private encryptionReadyPromise?: Promise<void>;
+    // A client that talks directly to the homeserver, bypassing pantalaimon.
+    private encryptionHsClient: MatrixClient;
+
+    constructor(
+        botSdkIntent: BotSdk.Intent,
+        botClient: BotSdk.MatrixClient,
+        intentOpts: IntentOpts, private encryptionOpts: EncryptedIntentOpts) {
+        super(botSdkIntent, botClient, intentOpts);
+
+        // We still need a direct client to the homeserver in some cases, so clone
+        // the existing one.
+        this.encryptionHsClient = new MatrixClient(
+            this.encryptionOpts.originalHomeserverUrl,
+            this.botSdkIntent.underlyingClient.accessToken
+        );
+        this.encryptionHsClient.impersonateUserId(this.userId);
+    }
+
+    /**
+     * Upload a file to the homeserver.
+     * @param content The file contents
+     * @param opts Additional options for the upload.
+     * @returns A MXC URL pointing to the uploaded data.
+     */
+    public async uploadContent(content: Buffer|string|ReadStream, opts: FileUploadOpts = {}): Promise<string> {
+        await this.ensureRegistered();
+        // Media is encrypted, since we don't know the destination room assume this media will be encrypted.
+        await this.encryptionOpts.ensureClientSyncingCallback();
+        return super.uploadContent(content, opts);
+    }
+
+    public onEvent(event: {
+        type: string,
+        // eslint-disable-next-line camelcase
+        content: {membership: UserMembership, displayname?: string, avatar_url?: string, algorithm?: string},
+        // eslint-disable-next-line camelcase
+        state_key: unknown,
+        // eslint-disable-next-line camelcase
+        room_id: string
+    }): void {
+        super.onEvent(event);
+        if (event.type === "m.room.encryption" && typeof event.content.algorithm === "string") {
+            log.info(`Room ${event.room_id} enabled encryption (${event.content.algorithm})`);
+            this.encryptedRooms.set(event.room_id, event.content.algorithm);
+        }
+    }
+
+    private async loginForEncryptedClient() {
+        const userId: string = this.userId;
+        const res = await this.botSdkIntent.underlyingClient.doRequest(
+            "POST",
+            "/_matrix/client/r0/login",
+            undefined,
+            {
+                type: APPSERVICE_LOGIN_TYPE,
+                identifier: {
+                    type: "m.id.user",
+                    user: userId,
+                }
+            },
+        );
+        return {
+            accessToken: res.access_token as string,
+            deviceId: res.device_id as string,
+        };
+    }
+
+    public async ensureRegistered(forceRegister = false): Promise<"registered=true"|undefined> {
```

Comment:
What's the reason for it returning this awkward string rather than a true (or just void, since it rejects on failure anyway?)

Reply:
It's a thing that the old old sdk did and we've just carried it forward. I'm not sure if anyone actually expects it..